SD-1276, SD-1633: Variables card

garyb opened the pull request.

garyb commented:
Alright, so this was more painful than I expected, but it's done at last.

Now we can have variables cards in multiple decks, we needed some way of qualifying the variable names with the deck IDs, so that a variable named `x` can be used in two different places, for example.

Qualifying every variable name with deck IDs in the URL would have made for some pretty heinous URLs and we might even have run into problems with URL length, since deck IDs are UUIDs, and therefore 36 characters long.

My workaround for this was to encode the varmaps in a JSON record which is passed to the URL instead. It's perhaps not the greatest solution either, but it seemed pretty reasonable, since these URLs aren't going to be hand written anyway.

The resulting code that the embed card generates now looks something like this:

``` html
<script type="text/javascript">
(function() {
  var varMaps = {
    "15ad5794-7405-471e-8d23-c3b42c10060d": {
      "var1": "2016-07-08",
      "var2": "A",
      "varX": "B",
      "say": "what"
    },
    "5845c3ff-443a-43af-a8a7-8a268d807683": {
      "var1": "foo"
    }
  };
  var uri = "http://local.slamdata.com/slamdata/workspace.html#/test/Untitled+Workspace+4.slam/76f59c11-c8ca-45cc-a92c-6bf8a1ef6077/view/?varMaps=" + encodeURIComponent(JSON.stringify(varMaps)) + "";
  var iframe = "<iframe width=\"100%\" height=\"100%\" frameborder=\"0\" src=\"" + uri + "\"></iframe>";
  document.writeln(iframe);
})();
</script>
```

So you can see we end up with a pretty nice arrangement here at least, that should be pretty simple for users to edit as they need.

/cc @jdegoes

garyb commented:
Actually, a thought, maybe I should rename `varMaps` to just `vars` in the URL and user facing code here?

natefaubion commented on src/SlamData/Workspace/Deck/Dialog/Share/Model.purs:
```diff
@@ -1,7 +1,40 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+-}
+
```
Doubly-licensed :) What, are you trying to bump up your +s now?

garyb replied:
Huh, weird, I just ran the script. And definitely not! In fact, this is going to interfere with my whole "more deletions than additions thing" :smile:

jdegoes commented:
> Actually, a thought, maybe I should rename varMaps to just vars in the URL and user facing code here?

That's good. Both in the URL and the code snippet, `vars` is much shorter.

Ideally we'd use deck names, but I realize deck identifiers are more stable, and anyway, we're not really doing much with deck names right now.

The other thing it'd be great to change while you're in here is the method of inserting the embedded deck.  Currently, it's done through `document.writeln`, which makes it impossible to call this code dynamically.

I think we should change to something like the following:

```javascript
<!-- You can put this in the header if you want -->
<script type="text/javascript">
if (typeof SlamData === "undefined") {
  var SlamData = {};
  SlamData.embed = function(deckPath, deckId, vars) {
    var uri = "http://local.slamdata.com/slamdata/workspace.html#" + deckPath + "/" + deckId + "/view/?vars=" + encodeURIComponent(JSON.stringify(vars)) + "";
    var iframe = document.createElement('iframe');  
    iframe.width = "100%"; iframe.height = "100%"; iframe.frameborder = 0;
    iframe.src = uri;
    var deckElement = document.getElementById(deckId);
    if (deckElement != null) deckElement.appendChild(iframe);
  }
}
</script>
<!-- This is the DOM element that the deck will be embedded into -->
<div id="15ad5794-7405-471e-8d23-c3b42c10060d"></div>
<!-- This is the code that performs the deck insertion-->
<script type="text/javascript">
  SlamData.embed(
    "/test/Untitled+Workspace+4.slam", 
    "15ad5794-7405-471e-8d23-c3b42c10060d",
    // These are variables passed to the deck(s), you can change them: 
    {
      "15ad5794-7405-471e-8d23-c3b42c10060d": {
        "var1": "2016-07-08",
        "var2": "A",
        "varX": "B",
        "say": "what"
      },
      "5845c3ff-443a-43af-a8a7-8a268d807683": {
        "var1": "foo"
      }
    });
</script>
```

Or something like it. The goal is to make it easy for a developer to programmatically call the embed code from Javascript with their own user-defined variables.

garyb commented:
Yeah, that's a lot better! :+1: I did consider updating what we had to avoid the `document.write` but figured this PR had already been long enough in coming.

garyb commented:
How about this?

``` html
<!-- This is the generic SlamData embedding code. -->
<!-- You can put this in the header if you want, or save it in a .js file included in the document -->
<script type="text/javascript">
var slamdata = window.SlamData = window.SlamData || {};
slamdata.embed = function(deckPath, deckId, vars) {
  var uri = "http://local.slamdata.com/slamdata/workspace.html#" + deckPath + "/" + deckId + "/view/?vars=" + encodeURIComponent(JSON.stringify(vars));
  var iframe = document.createElement('iframe');
  iframe.width = "100%"; iframe.height = "100%"; iframe.frameBorder = 0;
  iframe.src = uri;
  var deckElement = document.getElementById(deckId);
  if (deckElement !== null) deckElement.appendChild(iframe);
};
</script>

<!-- This is the DOM element that the deck will be embedded into -->
<div id="sd-deck-15ad5794-7405-471e-8d23-c3b42c10060d"></div>

<!-- This is the code that performs the deck insertion, placing it at the end of the body is suggested -->
<script type="text/javascript">
  SlamData.embed(
    "/test/Untitled+Workspace+4.slam",
    "sd-deck-15ad5794-7405-471e-8d23-c3b42c10060d",
    // These are variables passed to the deck(s), you can change them:
    {
      "15ad5794-7405-471e-8d23-c3b42c10060d": {
        "var1": "2016-07-08",
        "var2": "A",
        "varX": "B",
        "say": "what"
      },
      "5845c3ff-443a-43af-a8a7-8a268d807683": {
        "var1": "foo"
      }
    });
</script>
```

This way we extend or create a global SlamData object instead. Also added a suggestion that the embedding goes at the end of the body, so it will trigger the load once the rest of the document is ready.

jdegoes commented:
> This way we extend or create a global SlamData object instead. Also added a suggestion that the embedding goes at the end of the body, so it will trigger the load once the rest of the document is ready.

This looks really good. In the interest of making this even more friendly to developers, I'd suggest the following two additions:

1. Add a parameter for stylesheet URLs, as I know we support adding custom stylesheets via the embedded iframe URL. By default, this could be an empty array `[]`, indicating no extra stylesheets are added.
2. Add a parameter for permission tokens, since I know the embedded iframe URL supports such tokens. By default, this could be an empty array `[]`.

This way, a developer who gets this code knows exactly what they have to do to make it 100% operational inside their MongoDB application for the embedded reporting use case.

garyb commented:
Ok, great, will do! Point 2 I had realised already, and I'm looking at now. I'll probably change it so `embed` expects an options object rather than a list of arguments, so that way we can easily handle the various combinations of optional arguments by allowing them to be omitted entirely too.

jdegoes commented:
> Ok, great, will do! Point 2 I had realised already, and I'm looking at now. I'll probably change it so embed expects an options object rather than a list of arguments, so that way we can easily handle the various combinations of optional arguments by allowing them to be omitted entirely too.

👍

garyb commented:
Ok, latest version of the embed output:

``` html
<!-- This is the generic SlamData embedding code. -->
<!-- You can put this in the header or save it in a .js file included in the document -->
<script type="text/javascript">
var slamdata = window.SlamData = window.SlamData || {};
slamdata.embed = function(options) {
  var queryParts = [];
  if (options.tokenHash) queryParts.push("permissionToken=" + options.tokenHash);
  if (options.stylesheets) queryParts.push("stylesheets=" + options.stylesheets.map(encodeURIComponent).join(","));
  var queryString = "?" + queryParts.join("&");
  var varsParam = options.vars ? "/?vars=" + encodeURIComponent(JSON.stringify(options.vars)) : "";
  var uri = "http://local.slamdata.com/slamdata/workspace.html" + queryString;
  var iframe = document.createElement("iframe");
  iframe.width = iframe.height = "100%";
  iframe.frameBorder = 0;
  iframe.src = uri + "#" + options.deckPath + "/" + options.deckId + "/view" + varsParam;
  var deckElement = document.getElementById("sd-deck-" + options.deckId);
  if (deckElement) deckElement.appendChild(iframe);
};
</script>

<!-- This is the DOM element that the deck will be embedded into -->
<div id="sd-deck-76f59c11-c8ca-45cc-a92c-6bf8a1ef6077"></div>

<!-- This is the code that performs the deck insertion, placing it at the end of the body is suggested -->
<script type="text/javascript">
  SlamData.embed({
    deckPath: "/test/Untitled+Workspace+4.slam/",
    deckId: "76f59c11-c8ca-45cc-a92c-6bf8a1ef6077",
    // An array of custom stylesheets URLs can be provided here:
    stylesheets: [],
    // The variables for the deck(s), you can change their values here:
    vars: {
      "15ad5794-7405-471e-8d23-c3b42c10060d": {
        "var2": "A",
        "varX": "B",
        "say": "what",
        "var1": "2016-07-08T00:00:00Z"
      },
      "5845c3ff-443a-43af-a8a7-8a268d807683": {
        "var1": "foo"
      }
    }
  });
</script>
```

And `tokenHash` is included in the options if the user is logged in.

The generic embed code is kinda large now, it would be nice if we could provide it as an include-able .js file instead, and provide a script tag linking to it, but because the URL for the server is in that chunk of code we can't do that with a static js file, alas.

natefaubion commented on src/SlamData/Workspace/Deck/Dialog/Export/Component.purs:
```diff
@@ -14,520 +14,541 @@ See the License for the specific language governing permissions and
+      (do guard isLoggedIn
+          token ← permToken
+          token.secret)
+  ⊕ mkWorkspaceHash (deckPath' sharingInput.workspacePath sharingInput.deckId) (WA.Load AT.ReadOnly) Map.empty
```
This whole file has changed, is this a line ending issue?

garyb replied:
Ah yeah, it was CRLF for some reason.

garyb replied:
The only changes of note are in `renderCopyVal` which is pretty much entirely rewritten. Above there I think the only change is I renamed `isLogged` to `isLoggedIn`.

garyb replied:
And the type for `varMaps`.

natefaubion commented on src/SlamData/Workspace/Deck/Dialog/Export/Component.purs:
```diff
@@ -1,499 +1,554 @@
+      ⊕ line """    // An array of custom stylesheets URLs can be provided here"""
+      ⊕ line ("""    stylesheets: []""" ⊕ options)
+      ⊕ line """  });"""
+      ⊕ line """</script>"""
```
`Data.String.joinWith "\n"`?

garyb replied:
Good idea :smile:

natefaubion commented on src/SlamData/Workspace/Routing.purs:
```diff
@@ -133,30 +139,67 @@ mkWorkspaceURL
+  go (Port.QueryExpr q) =
+    -- | This is not entirely legit as it will strip backticks from SQL²
+    -- | expressions as well as identifiers, as we have no information about
+    -- | the field type here... -gb
```
Is this just an unlikely case or something?

garyb replied:
As far as I know, `` `something` `` is not a valid SQL² expression on its own, so I don't _think_ it's a problem. If we do need to fix this it's a bit of a pain, as we need to deal with `FormBuilder.Model.Item` rather than a `VarMap` here, and then I guess we need the `VarMap` when it's on the way in instead...

natefaubion replied:
I'll take your word for it.

natefaubion commented:
As long as @jdegoes approves of the embed API, then I think this is 👍

garyb commented:
Cool, thanks for reviewing! I'll wait for the @jdegoes signoff before merging.

jdegoes commented:
@garyb This looks almost perfect, except I remember from discussions with @cryogenian that "tokenHash" can actually be an array of tokens. Is that true?

cryogenian commented:
@jdegoes yes, it could be list of hashes separated by `,`

jdegoes commented:
@garyb @cryogenian In that case, let's call it plural and assume it's an array and join them together with `","`, just to make it easy for users to include multiple ones.

garyb commented on src/SlamData/Quasar/Auth/Permission.purs:
```diff
@@ -108,7 +108,7 @@ retrieveTokenHashes =
   where
   permissionRegex :: Rgx.Regex
-  permissionRegex = Rgx.regex "permissionsToken=([^&]+)" Rgx.noFlags
+  permissionRegex = Rgx.regex "permissionsTokens=([^&]+)" Rgx.noFlags
```
^ I assume this should have been the case too then?

garyb commented:
Okidoke, updated now, and renamed `tokenHash` to `permissionTokens` so it's less obscure too. I was a bit confused by the situation before, because the thing I just commented on expected `permissionToken` in the URL, so I assumed it was singular, but then the old embed code was generating a `permissionTokens` querystring... so it probably wasn't working either way :wink:

garyb commented:
@cryogenian can you take a look at just the last commit to confirm that's what we want?

cryogenian commented:
@garyb names differ. In one place it's `permissionsTokens` in other -- `permissionTokens`

garyb commented:
Good catch!

jdegoes commented:
Ideally, we should also do a manual test of this with Advanced with a generated permission token, to make sure there's nothing screwy about the URL and to make sure the final product looks correct when embedded. I assume @cryogenian's code for generating a permissions token in the publish URL has already been tested.

beckyconning commented:
@jdegoes that source hasn't been merged yet. can we merge this now then test integration with publish deck once its been reviewed and merged?

jdegoes commented:
@beckyconning Yep, works for me. 👍

garyb commented:
:tada:
